libass/ass_directwrite.c

```diff
@@ -726,7 +726,21 @@ static int CALLBACK font_enum_proc(const ENUMLOGFONTW *lpelf,
     if (FontType & RASTER_FONTTYPE)
         goto cleanup;
 
-    hFont = CreateFontIndirectW(&lpelf->elfLogFont);
+    LOGFONTW lf = lpelf->elfLogFont;
+    // lf.lfFaceName currently holds the font family name.
+    // The family may contain several fonts with equal numerical parameters
+    // but distinct full names. If we pass lf to CreateFontIndirect as is,
+    // we will miss this variety and merely get the same one font multiple
+    // times. Try to increase our chances of seeing all the right fonts
+    // by replacing the family name by the full name. There's some chance
+    // that this in turn selects some *other* family's fonts and misses
+    // the requested family, but we would rather take this chance than
+    // add every font twice (via the family name and via the full name).
+    // lfFaceName can hold up to LF_FACESIZE wchars; truncate longer names.
+    wcsncpy(lf.lfFaceName, lpelf->elfFullName, LF_FACESIZE - 1);
```

Comment:
It may be a bit cleaner to do ``wcsncpy_s(lf.lfFaceName, LF_FACESIZE, lpelf->elfFullName, _TRUNCATE);``

Reply:
> currently our baseline is ISO C99

This information doesn't seem to be written anywhere. It would be nice to have a contributing guide mentioning that.

> afaik all of this code currently still works/can be built for Windows XP

Technically, ``ass_directwrite.c`` is only for Windows Vista SP2 and later, as DWrite doesn't exist before that.

> Are there C11 compilers for Windows XP?

I don't know.

> building for Windows XP still relevant?

[VLC still support Windows XP](https://www.videolan.org/vlc/download-windows.html), so it is relevant.

Reply:
It being kinda nonstandard is part of the reason I don’t use it, but to be fair, this file is Windows-specific and it’s existed on Windows for (at least) almost two decades. It’s in Microsoft’s versioned Visual C++ libraries, it’s in UCRT, and it’s also in MSVCRT (so even compiling with 32-bit MinGW-w64 GCC works as I’ve just verified), but I don’t know when exactly it was added to MSVCRT and to MinGW headers.

Reply:
> > currently our baseline is ISO C99
>
> This information doesn't seem to be written anywhere. It would be nice to have a contributing guide mentioning that.

It’s what our buildsystem sets:

https://github.com/libass/libass/blob/e51781341493e0312f0c7fee66372f0365f7b962/Makefile.am#L3-L6

But admittedly this is not very prominent and buildflags also explicitly enabling GNU extensions is slightly misleading. We enable those since we want to make use of them if available but provide fallbacks if not available.

That being said, if Windows had this function for sufficiently long, using it in a Windows-specific file is absolutely fine.

Reply:
> Technically, `ass_directwrite.c` is only for Windows Vista SP2 and later, as DWrite doesn't exist before that.

Not quite. It cannot _run_ on Windows XP, but it can be built on Windows XP and run on Vista+, and binaries built on Vista+ can run on XP without using ass_directwrite at runtime.

Reply:
While the code with `_TRUNCATE` does look fairly nice, I feel I remain hesitant to use it for fear of it being absent in older/other compilers/libraries. So I’ll go ahead and merge this as is. But thanks for the suggestion; it may serve as a bit of education and inspiration for future code if/when we upgrade to newer language versions.